proto/schedule/utils.go

```diff
@@ -0,0 +1,76 @@
+package schedule
+
+import (
+	"fmt"
+	"github.com/ghodss/yaml"
+	"io/ioutil"
+	"time"
+)
+
+const (
+	DateFormat                  = "Mon 02 Jan 2006"
+	defaultStopTimeDurationDays = 7
+)
+
+func (s *Schedule) Validate() error {
+	if s == nil {
+		return fmt.Errorf("nil Schedule")
+	}
+
+	if s.Shifts == nil || len(s.Shifts) == 0 {
+		fmt.Printf("No input shifts found. Starting from scratch.")
+		return nil
+	}
+
+	for i, shift := range s.Shifts {
+		if _, err := time.Parse(DateFormat, shift.StartDate); err != nil {
+			return fmt.Errorf("error in input shift entry %v, invalid value: %v, err: %v", i, shift.StartDate, err)
+		}
+		if shift.User == "" {
+			return fmt.Errorf("user cannot be empty in input shift entry %v", i)
+		}
+	}
+
+	return nil // It's all good.
+}
+
+func (s *Schedule) FromYAMLFile(path string) error {
+	b, err := ioutil.ReadFile(path)
+	if err != nil {
+		return err
+	}
+
+	s.Reset()
+	return yaml.Unmarshal(b, s)
+}
+
+// Returns the time difference between the next-to-last and last Shifts in this Schedule as an estimation for when the
+// last Shift will end. Defaults to 7 when only 1 shift is in the schedule, and throws an error otherwise.
+func (s *Schedule) EstimateStopTime() (time.Time, error) {
+	if s == nil {
+		return time.Time{}, fmt.Errorf("no last shift on nil schedule")
+	}
+
+	if s.Shifts == nil || len(s.Shifts) == 0 {
+		return time.Time{}, fmt.Errorf("shifts cannot be nil or empty")
+	}
+
+	lastShift := s.Shifts[len(s.Shifts)-1]
```

Comment:
The need to do this estimation makes me wonder if we should change the data format we're using; should shifts have both a start date and an end date, instead of just assuming they stop at the next shift start date?  I guess that does denormalize data and could lead to inconsistencies, but that's effectively what the data will look like in the calendar anyway.

Reply:
Yeah, I've thought about changing the model over this, and thought a discussion would be nice.

One possible option would be to have a "stop" field that only the last shift would have. So something like:

```
shifts:
- startDate: Mon 01 Jun 2020
  user: abc
- startDate: Mon 08 Jun 2020
  user: lmn
- startDate: Mon 15 Jun 2020
  user: xyz
  stopDate: Sun 21 Jun 2020
```

Validate() could confirm a single `stopDate` on the last shift. WDYT? @plumpy @maggieneterval

Reply:
An alternate idea, which I mentioned (albeit post-merge) on the previous review was putting the shift length into the `Schedule` itself:
```
shiftDurationDays: 7
shifts:
  ...
```

This is admittedly putting configuration and data into the same file, but it feels okay to me. You could also include the username configuration in there too (a `oneof` for a static list of users _or_ a name of a GitHub team) so that the file contained both the schedule data and the configuration for how to generate additional schedule data.

I dunno, I'm sure it's controversial and happy to be talked down, but IMO it's the best approach for a tool like this.

Reply:
I thought about that with the post-merge comment, but ultimately it goes back to what Eric said above about the denormalization and inconsistency of the data - `shiftDurationDays` would become merely a suggestion.

The `stopDate` in the data conveys the same information in a consistent yet flexible way,